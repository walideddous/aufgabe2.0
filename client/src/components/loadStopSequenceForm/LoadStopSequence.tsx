import React, { useState, useCallback, useEffect } from "react";
import { AutoComplete, Card, Form, Select, Button } from "antd";
// Import types
import { TStopSequence } from "../../types/types";

interface TLoadStopSequence {
  show: boolean;
  currentMode: string[];
  stopSequenceList: TStopSequence[];
  currentStopSequence: TStopSequence | undefined;
  onStopsQuery: (modes: string[]) => void;
  onStopSequenceSearch: (name: string) => void;
  onDeleteStopSequence: (id: string) => void;
  onDisplayStopSequence: (modes: string[], key: string) => void;
}

const LoadStopSequence = ({
  show,
  currentMode,
  stopSequenceList,
  currentStopSequence,
  onStopsQuery,
  onStopSequenceSearch,
  onDeleteStopSequence,
  onDisplayStopSequence,
}: TLoadStopSequence) => {
  const [form] = Form.useForm();
  const [search, setSearch] = useState<string>("");
  const [selectValue, setSelectValue] = useState<string>(
    "verkehrsmittel typ auswählen"
  );

  // Auto complete component
  const { Option } = AutoComplete;

  useEffect(() => {
    if (currentMode[0]) {
      setSelectValue(currentMode[0] + "");
    }
  }, [currentMode]);

  // implementing the debouncing
  useEffect(() => {
    if (search) {
      const dispatchRequest = setTimeout(() => {
        onStopSequenceSearch(search);
      }, 1000);

      return () => clearTimeout(dispatchRequest);
    }
  }, [search, onStopSequenceSearch]);

  // handle the drop menu to display the choosed Modes on Map
  const handleModeChange = useCallback(
    (value: string) => {
      if (value !== "verkehrsmittel typ auswählen") {
        onStopsQuery([value + ""]);
        setSelectValue(value + "");
      }
    },
    [onStopsQuery]
  );

  const handleSelect = useCallback(
    (input: string, options) => {
      const response = stopSequenceList.filter(
        (el: any) => el.key === options.key
      )[0];
      if (response) {
        onDisplayStopSequence(response.modes, options.key);
      }
    },
    [stopSequenceList, onDisplayStopSequence]
  );

  return (
    <Card>
      <Form form={form} layout="vertical">
        {!show && (
          <Form.Item label="Verkehrsmittel typ">
            <Select
              id="mode_selector"
              value={selectValue}
              onChange={handleModeChange}
            >
              <Option value="verkehrsmittel typ auswählen" id="Modus_auswählen">
                verkehrsmittel typ auswählen
              </Option>
              <Option value="13" id="13">
                13
              </Option>
              <Option value="5" id="5">
                5
              </Option>
              <Option value="8" id="8">
                8
              </Option>
              <Option value="9" id="9">
                9
              </Option>
              <Option value="2" id="2">
                2
              </Option>
              <Option value="4" id="4">
                4
              </Option>
            </Select>
          </Form.Item>
        )}
        {show && (
          <>
            <Form.Item label="Haltestellensequenz name">
              <AutoComplete
                id="stopSequence_autoComplete"
                onChange={(input: string) => {
                  setSearch(input);
                }}
                value={search}
                onSelect={handleSelect}
                placeholder="Haltestellensequenz nach Namen suchen"
                allowClear={true}
              >
                {stopSequenceList &&
                  stopSequenceList.map((el: any) => (
                    <Option value={el.name} key={el.key}>
                      {el.name}
                    </Option>
                  ))}
              </AutoComplete>
            </Form.Item>
            {currentStopSequence && (
              <div style={{ display: "flex", justifyContent: "flex-end" }}>
                <Button
                  danger
                  id="delete_stopSequence"
                  disabled={
                    search &&
                    currentStopSequence &&
                    search === currentStopSequence.name
                      ? false
                      : true
                  }
                  onClick={() => {
                    if (currentStopSequence) {
                      const { key } = currentStopSequence;
                      onDeleteStopSequence(key);
                      setSearch("");
                    }
                  }}
                >
                  Haltestellensequenz löschen
                </Button>
              </div>
            )}
          </>
        )}
      </Form>
    </Card>
  );
};

export default React.memo(LoadStopSequence);
