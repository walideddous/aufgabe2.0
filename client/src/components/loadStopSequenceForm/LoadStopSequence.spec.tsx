import React from "react";
import { mount } from "enzyme";
import toJSON from "enzyme-to-json";
import LoadStopSequence from "./LoadStopSequence";

// import data to test
import {
  stopSequenceList,
  currentStopSequence,
} from "../../testUtils/testData";

const makeProps = (props: any) => ({
  stopSequenceList: [],
  currentStopSequence: {},
  loadMode() {},
  handleUpdateAfterSave() {},
  onStopsQuery() {},
  ondisplayStopSequence() {},
  handleDeleteStopSequenceMutation() {},
  onClearAll() {},
  ...props,
});

const setUp = (props: any) => {
  const component = mount(<LoadStopSequence {...props} />);
  return component;
};

describe("LoadStopSequence component", () => {
  let mountWrapper: any;

  const currentMode = ["4"];

  const onLoadMode = jest.fn();
  const onStopsQuery = jest.fn();
  const onUpdateAfterSave = jest.fn();
  const onClearAll = jest.fn();
  const onDeleteStopSequence = jest.fn();
  const ondisplayStopSequence = jest.fn();
  const onStopSequenceSearch = jest.fn();

  beforeEach(() => {
    mountWrapper = setUp(
      makeProps({
        stopSequenceList,
        currentStopSequence,
        currentMode,
        onStopSequenceSearch,
        onLoadMode,
        onStopsQuery,
        onClearAll,
        onDeleteStopSequence,
        ondisplayStopSequence,
      })
    );
  });

  afterEach(() => {
    jest.resetAllMocks();
  });

  it("Should test the jest framework", () => {
    expect(true).toBe(true);
  });

  it("Should match the LoadStopSequence component snapshot", () => {
    expect(toJSON(mountWrapper)).toMatchSnapshot();
  });

  it("Should disptach the onStopsQuery props function when we choose the mode", () => {
    console.log(mountWrapper);

    const radioButton = mountWrapper.find("#radioButton");

    radioButton.props().value = "new";

    const modeSelector = mountWrapper.find("#mode_selector");

    expect(modeSelector.props().value).toEqual("Choose mode");

    modeSelector.simulate("change", "4");

    expect(onStopsQuery).toHaveBeenCalledWith(["4"]);
  });

  it("Should dispatch the onClearAll props function when we choose the new button", () => {
    const radioButton = mountWrapper.find("#radioButton");

    console.log(radioButton);

    radioButton.simulate("change", {
      target: {
        value: "new",
      },
    });

    expect(onClearAll).toHaveBeenCalled();
    expect(onLoadMode).toHaveBeenCalledWith(false);
  });

  it("Should dispatch the handleUpdateAfterSave props function when we choose load button", () => {
    const radioButton = mountWrapper.find("#radioButton");

    radioButton.simulate("change", {
      target: {
        value: "load",
      },
    });

    expect(onUpdateAfterSave).toHaveBeenCalled();
    expect(onLoadMode).toHaveBeenCalledWith(true);
  });

  it("Should dispatch ondisplayStopSequence props function on select value in Auto-complete field ", () => {
    const radioButton = mountWrapper.find("#radioButton");

    radioButton.simulate("change", {
      target: {
        value: "load",
      },
    });

    const AutoCompleteInput = mountWrapper.find("#stopSequence_autoComplete");

    AutoCompleteInput.props().onChange("Test");
    AutoCompleteInput.props().onSelect("Test");

    expect(ondisplayStopSequence).toHaveBeenCalled();
  });

  it("Schould delete the stop sequence when we click on the delete stop sequence button", () => {
    const radioButton = mountWrapper.find("#radioButton");

    radioButton.simulate("change", {
      target: {
        value: "load",
      },
    });

    const AutoCompleteInput = mountWrapper.find("#stopSequence_autoComplete");

    AutoCompleteInput.props().onChange("Test");
    AutoCompleteInput.props().onSelect("Test");

    mountWrapper.find("#delete_stopSequence").at(0).simulate("click");

    expect(onDeleteStopSequence).toBeCalledWith(currentStopSequence._id);
  });
});
