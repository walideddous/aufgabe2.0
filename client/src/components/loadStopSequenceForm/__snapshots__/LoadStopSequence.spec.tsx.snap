// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoadStopSequence component Should match the LoadStopSequence component snapshot 1`] = `
<Card>
  <ForwardRef(InternalForm)
    form={
      Object {
        "__INTERNAL__": Object {
          "itemRef": [Function],
        },
        "getFieldError": [Function],
        "getFieldInstance": [Function],
        "getFieldValue": [Function],
        "getFieldsError": [Function],
        "getFieldsValue": [Function],
        "getInternalHooks": [Function],
        "isFieldTouched": [Function],
        "isFieldValidating": [Function],
        "isFieldsTouched": [Function],
        "isFieldsValidating": [Function],
        "resetFields": [Function],
        "scrollToField": [Function],
        "setFields": [Function],
        "setFieldsValue": [Function],
        "submit": [Function],
        "validateFields": [Function],
      }
    }
    layout="vertical"
  >
    <FormItem
      label="Verkehrsmittel typ"
    >
      <ForwardRef(InternalSelect)
        id="mode_selector"
        onChange={[Function]}
        value="verkehrsmittel typ ausw채hlen"
      >
        <Option
          id="Modus_ausw채hlen"
          value="verkehrsmittel typ ausw채hlen"
        >
          verkehrsmittel typ ausw채hlen
        </Option>
        <Option
          id="13"
          value="13"
        >
          13
        </Option>
        <Option
          id="5"
          value="5"
        >
          5
        </Option>
        <Option
          id="8"
          value="8"
        >
          8
        </Option>
        <Option
          id="9"
          value="9"
        >
          9
        </Option>
        <Option
          id="2"
          value="2"
        >
          2
        </Option>
        <Option
          id="4"
          value="4"
        >
          4
        </Option>
      </ForwardRef(InternalSelect)>
    </FormItem>
  </ForwardRef(InternalForm)>
</Card>
`;
